### ISSUE TITLE
Claims Not Coming with Access or Id Token

### ISSUE THREAD

**Issue by nosmalldreams** [IdentityServer/IdentityServer3#3095]
### Question / Issue
I am accessing IdentityServer using implicit flow from a JavaScript SPA application. I am using MembershipReboot to create users and to configure role claims for those users. I am logging in, requesting access and id tokens, from the SPA. The user I'm logging in with has a role claim associated, but I don't receive that role claim in the server response. Hoping you can help me figure out what I'm doing wrong.  

I've read this post - https://github.com/IdentityServer/IdentityServer3/issues/1784 - and my scope does specify both IncludeAllClaimsForUser, and each claim specifies alwaysInclude:true (I tried both hoping at least one would work). Here is the scope definition:
```
    new Scope
    {
        Name = "myAppScope",
        DisplayName = "My Application Scope",
        Type = ScopeType.Resource,
        IncludeAllClaimsForUser = true,
        Claims = new List<ScopeClaim>
        {
            new ScopeClaim("role", alwaysInclude:true), // alwaysInclude ensures roles are sent to client with access token
        }
     }
```
I'm definitely hitting the right scope - here is the value of the result received from the server:
![image](https://cloud.githubusercontent.com/assets/2496051/17156224/4c2fef10-534e-11e6-921b-5f25cfbb7f5c.png)

And here are the unencoded token contents, which seem to be the same for both the access and id tokens:
![image](https://cloud.githubusercontent.com/assets/2496051/17156269/894ad716-534e-11e6-8841-ceb252ef8fc1.png)

Note that the unencoded token contents do not contain any claim information. 

This is the code I'm using in the client application to request the tokens:
```
    public Authorize() {
        this.ResetAuthorizationData();

        //console.log("BEGIN Authorize, no auth data");

        var authorizationUrl = this.authorizationServerUrl + this.authorizationPath; // e.g. 'https://localhost:44328/connect/authorize';
        var client_id = this.authorizationClientId; // e.g. "jsClient"
        var redirect_uri = this.jsAppServerUrl + this.authorizationRedirectPath; // e.g. 'http://localhost:3000/home';
        // Request both identity token (for user info) and access token (for authorization). 
        // Note that to receive id_token, the scope must include 'openid'
        var response_type = "id_token token"; 
        var scope = "openid " + this.applicationAuthResourceScope;
        var nonce = "N" + Math.random() + "" + Date.now();
        var state = Date.now() + "" + Math.random();

        this.store("authStateControl", state);
        this.store("authNonce", nonce);
        //console.log("AuthorizedController created. adding myautostate: " + this.retrieve("authStateControl"));

        var url =
            authorizationUrl + "?" +
            "client_id=" + encodeURI(client_id) + "&" +
            "redirect_uri=" + encodeURI(redirect_uri) + "&" +
            "response_type=" + encodeURI(response_type) + "&" +
            "scope=" + encodeURI(scope) + "&" +
            "nonce=" + encodeURI(nonce) + "&" +
            "state=" + encodeURI(state);

        window.location.href = url;
    }
```

I am hoping very much that you may be able to help me understand what I am doing wrong / what I am misunderstanding. Many thanks!

- [ ] I read and understood how to enable [logging](https://identityserver.github.io/Documentation/docsv2/configuration/logging.html)

### Relevant parts of the log file

```
2016-07-26 16:38:06,934 [12] INFO  IdentityServer3.Core.Endpoints.AuthorizeEndpointController Start authorize request
2016-07-26 16:38:06,935 [12] INFO  IdentityServer3.Core.Validation.AuthorizeRequestValidator Start authorize request protocol validation
2016-07-26 16:38:06,936 [12] INFO  IdentityServer3.Core.Validation.AuthorizeRequestValidator Authorize request validation success
 {
  "ClientId": "jsClient",
  "ClientName": "JS Client",
  "RedirectUri": "http://localhost:3000/home",
  "AllowedRedirectUris": [
    "http://localhost:3000/home"
  ],
  "SubjectId": "unknown",
  "ResponseType": "id_token token",
  "ResponseMode": "fragment",
  "Flow": "Implicit",
  "RequestedScopes": "openid myAppScope",
  "State": "14695690869000.7958451534639011",
  "Nonce": "N0.40564013431378811469569086900",
  "Raw": {
    "client_id": "jsClient",
    "redirect_uri": "http://localhost:3000/home",
    "response_type": "id_token token",
    "scope": "openid myAppScope",
    "nonce": "N0.40564013431378811469569086900",
    "state": "14695690869000.7958451534639011"
  }
}
2016-07-26 16:38:06,937 [12] INFO  IdentityServer3.Core.ResponseHandling.AuthorizeInteractionResponseGenerator User is not authenticated. Redirecting to login.
2016-07-26 16:38:06,945 [12] INFO  IdentityServer3.Core.Endpoints.AuthorizeEndpointController End authorize request
2016-07-26 16:38:06,948 [12] INFO  IdentityServer3.Core.Results.LoginResult Redirecting to login page
2016-07-26 16:38:06,952 [12] DEBUG IdentityServer3.Core.Configuration.Hosting.MessageCookie`1 Protecting message: {"ReturnUrl":"https://localhost:44328/core/connect/authorize?client_id=jsClient&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fhome&response_type=id_token%20token&scope=openid%20myAppScope&nonce=N0.40564013431378811469569086900&state=14695690869000.7958451534639011","ClientId":"jsClient","AcrValues":[],"Created":636051658869339398}
2016-07-26 16:38:06,965 [15] INFO  IdentityServer3.Core.Endpoints.AuthenticationController Login page requested
2016-07-26 16:38:06,967 [15] DEBUG IdentityServer3.Core.Endpoints.AuthenticationController signin message passed to login: {
  "ReturnUrl": "https://localhost:44328/core/connect/authorize?client_id=jsClient&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fhome&response_type=id_token%20token&scope=openid%20myAppScope&nonce=N0.40564013431378811469569086900&state=14695690869000.7958451534639011",
  "ClientId": "jsClient",
  "IdP": null,
  "Tenant": null,
  "LoginHint": null,
  "DisplayMode": null,
  "UiLocales": null,
  "AcrValues": [],
  "Created": 636051658869339398
}
2016-07-26 16:38:06,975 [15] INFO  IdentityServer3.Core.Endpoints.AuthenticationController rendering login page
2016-07-26 16:38:27,350 [12] INFO  IdentityServer3.Core.Endpoints.AuthenticationController Login page submitted
2016-07-26 16:38:27,824 [12] INFO  IdentityServer3.Core.Endpoints.AuthenticationController Login credentials successfully validated by user service
2016-07-26 16:38:27,830 [12] INFO  IdentityServer3.Core.Endpoints.AuthenticationController Calling PostAuthenticateAsync on the user service
2016-07-26 16:38:27,832 [12] INFO  IdentityServer3.Core.Endpoints.AuthenticationController issuing primary signin cookie
2016-07-26 16:38:27,834 [12] INFO  IdentityServer3.Core.Endpoints.AuthenticationController redirecting to: https://localhost:44328/core/connect/authorize?client_id=jsClient&redirect_uri=http:%2F%2Flocalhost:3000%2Fhome&response_type=id_token token&scope=openid myAppScope&nonce=N0.40564013431378811469569086900&state=14695690869000.7958451534639011
2016-07-26 16:38:27,850 [22] INFO  IdentityServer3.Core.Endpoints.AuthorizeEndpointController Start authorize request
2016-07-26 16:38:27,851 [22] INFO  IdentityServer3.Core.Validation.AuthorizeRequestValidator Start authorize request protocol validation
2016-07-26 16:38:27,852 [22] INFO  IdentityServer3.Core.Validation.AuthorizeRequestValidator Authorize request validation success
 {
  "ClientId": "jsClient",
  "ClientName": "JS Client",
  "RedirectUri": "http://localhost:3000/home",
  "AllowedRedirectUris": [
    "http://localhost:3000/home"
  ],
  "SubjectId": "d428af74-efdf-40f4-aea5-93063874a805",
  "ResponseType": "id_token token",
  "ResponseMode": "fragment",
  "Flow": "Implicit",
  "RequestedScopes": "openid myAppScope",
  "State": "14695690869000.7958451534639011",
  "Nonce": "N0.40564013431378811469569086900",
  "SessionId": "c71dc888166d88808ba54a65f30ffbed",
  "Raw": {
    "client_id": "jsClient",
    "redirect_uri": "http://localhost:3000/home",
    "response_type": "id_token token",
    "scope": "openid myAppScope",
    "nonce": "N0.40564013431378811469569086900",
    "state": "14695690869000.7958451534639011"
  }
}
2016-07-26 16:38:27,872 [22] INFO  IdentityServer3.Core.ResponseHandling.AuthorizeResponseGenerator Creating Implicit Flow response.
2016-07-26 16:38:27,872 [22] DEBUG IdentityServer3.Core.Services.Default.DefaultTokenService Creating access token
2016-07-26 16:38:27,929 [22] DEBUG IdentityServer3.Core.Services.Default.DefaultTokenService Creating JWT access token
2016-07-26 16:38:27,938 [22] DEBUG IdentityServer3.Core.Services.Default.DefaultTokenService Creating identity token
2016-07-26 16:38:27,939 [22] INFO  IdentityServer3.Core.Services.Default.DefaultClaimsProvider Getting claims for identity token for subject: d428af74-efdf-40f4-aea5-93063874a805
2016-07-26 16:38:27,973 [22] DEBUG IdentityServer3.Core.Services.Default.DefaultTokenService Creating JWT identity token
2016-07-26 16:38:27,982 [22] DEBUG IdentityServer3.Core.Endpoints.AuthorizeEndpointController Adding client jsClient to client list cookie for subject d428af74-efdf-40f4-aea5-93063874a805
2016-07-26 16:38:27,983 [22] INFO  IdentityServer3.Core.Endpoints.AuthorizeEndpointController End authorize request
2016-07-26 16:38:27,984 [22] INFO  IdentityServer3.Core.Results.AuthorizeRedirectResult Redirecting to: http://localhost:3000/home

```

**Comment by leastprivilege** [comment]
put a breakpoint into GetProfileDataAsync and you will see what's going on.

**Comment by nosmalldreams** [comment]
Thank you for your quick response. I have set the breakpoint, and I see that the ProfileDataRequestContext.RequestedClaimTypes is coming through as null, and that's why the roles claims aren't getting returned. However, I still don't understand why that's the case / what I should be doing differently. 

Per [this](https://github.com/IdentityServer/IdentityServer3/issues/1621) post, it sounds like id_token is the correct response type for me to be requesting, and based on both that post and the documentation for scope definition [here](https://identityserver.github.io/Documentation/docsv2/configuration/scopesAndClaims.html), it sounds like the "alwaysInclude:true" option I've included in the claim should trigger it to get included in the id token. I must be missing something, but can't seem to figure out what...
